src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs

```diff
@@ -133,59 +171,67 @@ public bool ExecuteCommand(ExtractMethodCommandArgs args, CommandExecutionContex
 
-            // apply the change to buffer
-            // get method name token
-            SyntaxToken methodNameAtInvocation;
-            using (var undoTransaction = _undoManager.GetTextBufferUndoManager(textBuffer).TextBufferUndoHistory.CreateTransaction("Extract Method"))
-            {
-                // apply extract method code to buffer
-                (var formattedDocument, methodNameAtInvocation) = result.GetFormattedDocumentAsync(cleanupOptions, cancellationToken).WaitAndGetResult(cancellationToken);
-                formattedDocument.Project.Solution.Workspace.ApplyDocumentChanges(formattedDocument, cancellationToken);
-
-                // apply changes
-                undoTransaction.Complete();
-            }
+            await ApplyChangeAsync(formattedDocument, textBuffer, waitContext).ConfigureAwait(false);
 
             // start inline rename
-            var snapshotAfterFormatting = textBuffer.CurrentSnapshot;
-            var documentAfterFormatting = snapshotAfterFormatting.GetOpenDocumentInCurrentContextWithChanges();
-            _renameService.StartInlineSession(documentAfterFormatting, methodNameAtInvocation.Span, cancellationToken);
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var textSnapshot = textBuffer.CurrentSnapshot;
+            document = textSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document != null)
+                _renameService.StartInlineSession(document, methodNameAtInvocation.Span, cancellationToken);
 
             // select invocation span
-            view.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(snapshotAfterFormatting, methodNameAtInvocation.Span.End));
-            view.SetSelection(
-                methodNameAtInvocation.Span.ToSnapshotSpan(snapshotAfterFormatting));
+            view.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(textSnapshot, methodNameAtInvocation.Span.End));
+            view.SetSelection(methodNameAtInvocation.Span.ToSnapshotSpan(textSnapshot));
+        }
 
-            return true;
+        private async Task ApplyChangeAsync(
+            Document formattedDocument, ITextBuffer textBuffer, IBackgroundWorkIndicatorContext waitContext)
+        {
+            var cancellationToken = waitContext.UserCancellationToken;
+
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            using var undoTransaction = _undoManager.GetTextBufferUndoManager(textBuffer).TextBufferUndoHistory.CreateTransaction("Extract Method");
+
+            // We're about to make an edit ourselves.  so disable the cancellation that happens on editing.
+            waitContext.CancelOnEdit = false;
+            formattedDocument.Project.Solution.Workspace.ApplyDocumentChanges(formattedDocument, cancellationToken);
```

Comment:
1.  what happens if we cancel in the middle of the undo transaction?  Do we need to explicitly complete it / throw it away if we cancel?
2.  Is it possible to partially apply changes in ApplyDocumentChanges?  Or if it cancels does it do all or nothing

Reply:
so we're on the UI thread now.  so it shouldn't be possible to have a cancel happen at this point since only UI affinitized ops can cause cancellation.  it's a bit of an odd duck here i admit as we could also effectively have the same effect by passing CancellationToken.None here :-/

Reply:
interesting - but we still need the cancelOnEdit = false so that it doesn't immediately cancel after we finish applying changes, switch off the UI thread and cancel?

Reply:
yes,  we still need the cancelOnEdit false.  but not because of any switching.  because if we don't, we'll make the edit, and the cancellation token will get signaled.  ANd then any further work we do (on any thread) that checks the token will think we are cancled and will abort.